File: algorithms/2_medium/F/01371_find_the_longest_substring_containing_vowels_in_even_counts/solution.c
#define INVALID PTRDIFF_MIN

typedef ptrdiff_t isize;

inline static size_t get_mask_char(const char ch) {
    switch (ch) {
        case 'a': return 1;
        case 'e': return (1 << 1);
        case 'i': return (1 << 2);
        case 'o': return (1 << 3);
        case 'u': return (1 << 4);
        default: return 0;
    }
}

int findTheLongestSubstring(const char *restrict s) {
    isize first_seen[32];
    for (isize i = 0; i < 32; i += 1) { first_seen[i] = INVALID; }
    first_seen[0] = -1;
    size_t curr_state = 0;
    isize longest = 0;

    for (isize i = 0; s[i] != '\0'; i += 1) {
        const size_t mask = get_mask_char(s[i]);
        curr_state ^= mask;
        if (first_seen[curr_state] != INVALID) {
            longest = MAX(longest, i - first_seen[curr_state]);
        } else {
            first_seen[curr_state] = i;
        }
    }

    return (int)longest;
}

File: algorithms/2_medium/F/01371_find_the_longest_substring_containing_vowels_in_even_counts/solution.rs
impl Solution {
    #[inline(always)]
    fn get_mask_byte(b: u8) -> usize {
        match b {
            b'a' => 1,
            b'e' => (1 << 1),
            b'i' => (1 << 2),
            b'o' => (1 << 3),
            b'u' => (1 << 4),
            _ => 0,
        }
    }

    pub fn find_the_longest_substring(s: String) -> i32 {
        let mut first_seen: [Option<isize>; 32] = [None; 32];
        first_seen[0] = Some(-1);
        let mut curr_state = 0_usize;
        let mut longest = 0_isize;

        for (i, &b) in s.as_bytes().iter().enumerate() {
            let mask = Self::get_mask_byte(b);
            curr_state ^= mask;
            if let Some(state) = first_seen[curr_state] {
                longest = longest.max((i as isize) - state);
            } else {
                first_seen[curr_state] = Some(i as isize);
            }
        }

        longest.try_into().unwrap()
    }
}
